File: ycs-golang/core/abstract_type.go
package core

import (
	"ycs/content"
	"ycs/contracts"
)

// AbstractType represents the base type for all Y types
type AbstractType struct {
	item             contracts.IStructItem
	start            contracts.IStructItem
	m                map[string]contracts.IStructItem
	doc              contracts.IYDoc
	length           int
	eventHandler     func(contracts.YEventArgs)
	deepEventHandler func(contracts.YDeepEventArgs)
}

// NewAbstractType creates a new AbstractType
func NewAbstractType() *AbstractType {
	return &AbstractType{
		m: make(map[string]contracts.IStructItem),
	}
}

// GetItem returns the item
func (at *AbstractType) GetItem() contracts.IStructItem {
	return at.item
}

// SetItem sets the item
func (at *AbstractType) SetItem(item contracts.IStructItem) {
	at.item = item
}

// GetStart returns the start item
func (at *AbstractType) GetStart() contracts.IStructItem {
	return at.start
}

// SetStart sets the start item
func (at *AbstractType) SetStart(start contracts.IStructItem) {
	at.start = start
}

// GetMap returns the map
func (at *AbstractType) GetMap() map[string]contracts.IStructItem {
	return at.m
}

// SetMap sets the map
func (at *AbstractType) SetMap(m map[string]contracts.IStructItem) {
	at.m = m
}

// GetDoc returns the document
func (at *AbstractType) GetDoc() contracts.IYDoc {
	return at.doc
}

// GetParent returns the parent
func (at *AbstractType) GetParent() contracts.IAbstractType {
	if at.item != nil {
		return at.item.GetParent().(contracts.IAbstractType)
	}
	return nil
}

// GetLength returns the length
func (at *AbstractType) GetLength() int {
	return at.length
}

// SetLength sets the length
func (at *AbstractType) SetLength(length int) {
	at.length = length
}

// Integrate integrates the type with a document and item
func (at *AbstractType) Integrate(doc contracts.IYDoc, item contracts.IStructItem) {
	at.doc = doc
	at.item = item
}

// InternalCopy creates an internal copy (to be overridden by subclasses)
func (at *AbstractType) InternalCopy() contracts.IAbstractType {
	panic("InternalCopy not implemented")
}

// InternalClone creates an internal clone (to be overridden by subclasses)
func (at *AbstractType) InternalClone() contracts.IAbstractType {
	panic("InternalClone not implemented")
}

// Write writes the type to an encoder (to be overridden by subclasses)
func (at *AbstractType) Write(encoder contracts.IUpdateEncoder) {
	panic("Write not implemented")
}

// CallTypeObservers calls event listeners with an event. This will also add an event to all parents
// for observeDeep handlers.
func (at *AbstractType) CallTypeObservers(transaction contracts.ITransaction, evt contracts.IYEvent) {
	currentType := at

	for {
		values, exists := transaction.GetChangedParentTypes()[currentType]
		if !exists {
			values = make([]contracts.IYEvent, 0)
			transaction.GetChangedParentTypes()[currentType] = values
		}

		transaction.GetChangedParentTypes()[currentType] = append(values, evt)

		if currentType.item == nil {
			break
		}

		currentType = currentType.item.GetParent().(*AbstractType)
	}

	at.InvokeEventHandlers(evt, transaction)
}

// CallObserver creates YEvent and calls all type observers.
// Must be implemented by each type.
func (at *AbstractType) CallObserver(transaction contracts.ITransaction, parentSubs map[string]struct{}) {
	// Default implementation does nothing
}

// First returns the first non-deleted item
func (at *AbstractType) First() contracts.IStructItem {
	n := at.start
	for n != nil && n.GetDeleted() {
		n = n.GetRight()
	}
	return n
}

// InvokeEventHandlers invokes event handlers
func (at *AbstractType) InvokeEventHandlers(evt contracts.IYEvent, transaction contracts.ITransaction) {
	if at.eventHandler != nil {
		at.eventHandler(contracts.YEventArgs{Event: evt, Transaction: transaction})
	}
}

// CallDeepEventHandlerListeners calls deep event handler listeners
func (at *AbstractType) CallDeepEventHandlerListeners(events []contracts.IYEvent, transaction contracts.ITransaction) {
	if at.deepEventHandler != nil {
		at.deepEventHandler(contracts.YDeepEventArgs{Events: events, Transaction: transaction})
	}
}

// FindRootTypeKey finds the root type key
func (at *AbstractType) FindRootTypeKey() string {
	return at.doc.FindRootTypeKey(at)
}

// typeMapDelete deletes a key from the type map
func (at *AbstractType) typeMapDelete(transaction contracts.ITransaction, key string) {
	if c, exists := at.m[key]; exists {
		c.Delete(transaction)
	}
}

// typeMapSet sets a value in the type map
func (at *AbstractType) typeMapSet(transaction contracts.ITransaction, key string, value interface{}) {
	var left contracts.IStructItem
	if l, exists := at.m[key]; exists {
		left = l
	}

	doc := transaction.GetDoc()
	ownClientID := doc.GetClientID()
	contentFactory := content.MustGetGlobalFactory()
	contentObj := contentFactory.CreateContent(value)

	newItem := NewStructItem(
		StructID{Client: int64(ownClientID), Clock: doc.GetStore().GetState(int64(ownClientID))},
		left,
		FromContractsStructID(left.GetLastID()).ToPointer(),
		nil,
		nil,
		at,
		&key,
		contentObj,
	)
	newItem.Integrate(transaction, 0)
}

// tryTypeMapGet tries to get a value from the type map
func (at *AbstractType) tryTypeMapGet(key string) (interface{}, bool) {
	if val, exists := at.m[key]; exists && !val.GetDeleted() {
		content := val.GetContent().GetContent()
		return content[val.GetLength()-1], true
	}
	return nil, false
}

// typeMapGetSnapshot gets a value from the type map at a specific snapshot
func (at *AbstractType) typeMapGetSnapshot(key string, snapshot contracts.ISnapshot) interface{} {
	var v contracts.IStructItem
	if val, exists := at.m[key]; exists {
		v = val
	}

	stateVector := snapshot.GetStateVector()
	for v != nil {
		clientClock, hasClient := stateVector[v.GetID().Client]
		if !hasClient || v.GetID().Clock >= clientClock {
			v = v.GetLeft()
		} else {
			break
		}
	}

	if v != nil && v.IsVisible(snapshot) {
		content := v.GetContent().GetContent()
		return content[v.GetLength()-1]
	}
	return nil
}

// typeMapEnumerate enumerates non-deleted items in the type map
func (at *AbstractType) typeMapEnumerate() map[string]contracts.IStructItem {
	result := make(map[string]contracts.IStructItem)
	for key, value := range at.m {
		if !value.GetDeleted() {
			result[key] = value
		}
	}
	return result
}

// typeMapEnumerateValues enumerates values in the type map
func (at *AbstractType) typeMapEnumerateValues() map[string]interface{} {
	result := make(map[string]interface{})
	for key, value := range at.typeMapEnumerate() {
		content := value.GetContent().GetContent()
		result[key] = content[value.GetLength()-1]
	}
	return result
}

File: src/Ycs/Core/AbstractType.cs
// ------------------------------------------------------------------------------
//  <copyright company="Microsoft Corporation">
//      Copyright (c) Microsoft Corporation.  All rights reserved.
//  </copyright>
// ------------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Linq;
using Ycs.Contracts;
using Ycs.Types;
using Ycs.Content;

namespace Ycs.Core
{
    public class YEventArgs
    {
        public YEventArgs(IYEvent evt, ITransaction transaction)
        {
            Event = evt;
            Transaction = transaction;
        }

        public IYEvent Event { get; }
        public ITransaction Transaction { get; }
    }

    public class YDeepEventArgs
    {
        public YDeepEventArgs(IList<IYEvent> events, ITransaction transaction)
        {
            Events = events;
            Transaction = transaction;
        }

        public IList<IYEvent> Events { get; }
        public ITransaction Transaction { get; }
    }

    public class AbstractType : IAbstractType
    {
        public IItem Item { get; set; }
        public IItem Start { get; set; }
        public IDictionary<string, IItem> Map { get; set; } = new Dictionary<string, IItem>();

        public event EventHandler<YEventArgs> EventHandler;
        public event EventHandler<YDeepEventArgs> DeepEventHandler;

        public IYDoc Doc { get; protected set; }
        public IAbstractType Parent => Item != null ? Item.Parent as AbstractType : null;

        public virtual int Length { get; set; }

        public virtual void Integrate(IYDoc doc, IItem item)
        {
            Doc = doc;
            Item = item;
        }
        public virtual IAbstractType InternalCopy() { throw new NotImplementedException(); }
        public virtual IAbstractType InternalClone() { throw new NotImplementedException(); }

        public virtual void Write(IUpdateEncoder encoder) { throw new NotImplementedException(); }

        /// <summary>
        /// Call event listeners with an event. This will also add an event to all parents
        /// for observeDeep handlers.
        /// </summary>
        public virtual void CallTypeObservers(ITransaction transaction, IYEvent evt)
        {
            var type = this;

            while (true)
            {
                if (!transaction.ChangedParentTypes.TryGetValue(type, out var values))
                {
                    values = new List<IYEvent>();
                    transaction.ChangedParentTypes[type] = values;
                }

                values.Add(evt);

                if (type.Item == null)
                {
                    break;
                }

                type = type.Item.Parent as AbstractType;
            }

            InvokeEventHandlers(evt, transaction);
        }

        /// <summary>
        /// Creates YEvent and calls all type observers.
        /// Must be implemented by each type.
        /// </summary>
        public virtual void CallObserver(ITransaction transaction, ISet<string> parentSubs)
        {
            // Do nothing.
        }

        public IItem _First()
        {
            var n = Start;
            while (n != null && n.Deleted)
            {
                n = n.Right as IItem;
            }
            return n;
        }

        public void InvokeEventHandlers(IYEvent evt, ITransaction transaction)
        {
            EventHandler?.Invoke(this, new YEventArgs(evt, transaction));
        }

        public void CallDeepEventHandlerListeners(IList<IYEvent> events, ITransaction transaction)
        {
            DeepEventHandler?.Invoke(this, new YDeepEventArgs(events, transaction));
        }

        public string FindRootTypeKey()
        {
            return Doc.FindRootTypeKey(this);
        }

        protected void TypeMapDelete(ITransaction transaction, string key)
        {
            if (Map.TryGetValue(key, out var c))
            {
                c.Delete(transaction);
            }
        }

        protected void TypeMapSet(ITransaction transaction, string key, object value)
        {
            if (!Map.TryGetValue(key, out var left))
            {
                left = null;
            }

            var doc = transaction.Doc;
            var ownClientId = doc.ClientId;
            IContent content;

            if (value == null)
            {
                content = new ContentAny(new object[] { value });
            }
            else
            {
                switch (value)
                {
                    case YDoc d:
                        content = new ContentDoc(d);
                        break;
                    case AbstractType at:
                        content = new ContentType(at);
                        break;
                    case byte[] ba:
                        content = new ContentBinary(ba);
                        break;
                    default:
                        content = new ContentAny(new[] { value });
                        break;
                }
            }

            var newItem = new StructItem(new StructID(ownClientId, doc.Store.GetState(ownClientId)), left, left?.LastId, null, null, this, key, content);
            newItem.Integrate(transaction, 0);
        }

        protected bool TryTypeMapGet(string key, out object value)
        {
            if (Map.TryGetValue(key, out var val) && !val.Deleted)
            {
                value = val.Content.GetContent()[val.Length - 1];
                return true;
            }

            value = default;
            return false;
        }

        protected object TypeMapGetSnapshot(string key, ISnapshot snapshot)
        {
            if (!Map.TryGetValue(key, out var v))
            {
                v = null;
            }

            while (v != null && (!snapshot.StateVector.ContainsKey(v.Id.Client) || v.Id.Clock >= snapshot.StateVector[v.Id.Client]))
            {
                v = v.Left as IItem;
            }

            return v != null && v.IsVisible(snapshot) ? v.Content.GetContent()[v.Length - 1] : null;
        }

        protected IEnumerable<KeyValuePair<string, IItem>> TypeMapEnumerate() => Map.Where(kvp => !kvp.Value.Deleted);

        protected IEnumerable<KeyValuePair<string, object>> TypeMapEnumerateValues()
        {
            foreach (var kvp in TypeMapEnumerate())
            {
                var key = kvp.Key;
                var value = kvp.Value.Content.GetContent()[kvp.Value.Length - 1];
                yield return new KeyValuePair<string, object>(key, value);
            }
        }
    }
}
